compiler/rustc_trait_selection/src/error_reporting/infer/mod.rs

```diff
@@ -1836,9 +1836,10 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
             if should_suggest_fixes
                 && !matches!(terr, TypeError::RegionsInsufficientlyPolymorphic(..))
             {
+                let hir = self.tcx.hir_node_by_def_id(cause.body_id);
```

Comment:
push this down into `suggest_await_on_expected_found`, no need to separate the hir access from the place it's actually used

Reply:
Actually, could you match on `self.tcx.coroutine_kind(self.body_id)`? Perhaps you should share the logic with https://github.com/rust-lang/rust/pull/131549.

Reply:
But anyways, make sure that check is inside of `suggest_await_on_expected_found `.